Sources/Core/X86/HardwareBreakpointManager.cpp

```diff
@@ -260,4 +242,35 @@ ErrorCode HardwareBreakpointManager::readDebugRegisters(
 
   return kSuccess;
 }
+
+ErrorCode HardwareBreakpointManager::writeDebugRegisters(
+    Target::Thread *thread, std::vector<uint64_t> &regs) const {
+  Architecture::CPUState state;
+
+  CHK(thread->readCPUState(state));
+
+#if defined(ARCH_X86)
+  for (int i = 0; i < kNumDebugRegisters; ++i) {
+    state.dr.dr[i] = (i == 4 || i == 5) ? 0 : regs[i];
+  }
+#elif defined(ARCH_X86_64)
+  for (int i = 0; i < kNumDebugRegisters; ++i) {
+    if (state.is32) {
+      state.state32.dr.dr[i] = (i == 4 || i == 5) ? 0 : regs[i];
+    } else {
+      if (i < 4) {
+        state.state64.dr.dr[i] = regs[i];
+      } else if (i == 4 || i == 5) {
+        state.state64.dr.dr[i] = 0;
+      } else {
+        state.state64.dr.dr[i] = regs[i] & 0xffffffff;
```

Comment:
Same comment here. This should be handled in `CPUState` directly to centralize the logic. After, this function can become simpler as well.